File: safe-online-sdk/src/main/java/net/link/safeonline/sdk/ws/data/DataClient.java
/*
 * SafeOnline project.
 *
 * Copyright 2006-2007 Lin.k N.V. All rights reserved.
 * Lin.k N.V. proprietary/confidential. Use is subject to license terms.
 */

package net.link.safeonline.sdk.ws.data;

import net.link.safeonline.sdk.exception.AttributeNotFoundException;
import net.link.safeonline.sdk.exception.RequestDeniedException;
import net.link.safeonline.sdk.exception.SubjectNotFoundException;
import net.link.safeonline.sdk.ws.MessageAccessor;
import net.link.safeonline.sdk.ws.exception.WSClientTransportException;


/**
 * Interface for data client component. Via this interface application can perform CRUD operations on attributes of a subject. For this the
 * application must be an attribute provider. Only the operator can set the attribute provider role for applications.
 * 
 * @author fcorneli
 * 
 */
public interface DataClient extends MessageAccessor {

    /**
     * Sets the value of an attribute. Please notice that the attribute should already be defined via:
     * {@link #createAttribute(String, String, Object)}. The attribute value can be of type {@link String} or {@link Boolean}.
     * 
     * @param userId
     * @param attributeName
     * @param attributeValue
     * @throws WSClientTransportException
     *             in case the service could not be contacted. Can happen if the SSL was not setup correctly.
     * @throws AttributeNotFoundException
     *             in case the attribute entity did not exist.
     * @see #createAttribute(String, String, Object)
     */
    void setAttributeValue(String userId, String attributeName, Object attributeValue) throws WSClientTransportException,
                                                                                      AttributeNotFoundException;

    /**
     * Gives back the attribute value of an attribute. We return an {@link Attribute} object to be able to make a distinction between a
     * missing attribute and a <code>null</code> attribute value.
     * 
     * @param <Type>
     *            the type of the attribute value.
     * @param userId
     * @param attributeName
     * @param valueClass
     *            the type of the attribute value.
     * @throws WSClientTransportException
     *             in case the service could not be contacted. Can happen if the SSL was not setup correctly.
     * @throws RequestDeniedException
     * @throws SubjectNotFoundException
     */
    <Type> Attribute<Type> getAttributeValue(String userId, String attributeName, Class<Type> valueClass)
                                                                                                         throws WSClientTransportException,
                                                                                                         RequestDeniedException,
                                                                                                         SubjectNotFoundException;

    /**
     * Creates a new (empty) attribute for the given subject.
     * 
     * @param userId
     * @param attributeName
     * @throws WSClientTransportException
     *             in case the service could not be contacted. Can happen if the SSL was not setup correctly.
     */
    void createAttribute(String userId, String attributeName, Object objectValue) throws WSClientTransportException;

    /**
     * Removes an attribute for the given subject.
     * 
     * @param userId
     *            the subject from which to remove the attribute.
     * @param attributeName
     *            the name of the attribute to be removed.
     * @param attributeId
     *            the optional attributeId in case of a compounded attribute.
     * @throws WSClientTransportException
     *             in case the service could not be contacted. Can happen if the SSL was not setup correctly.
     */
    void removeAttribute(String userId, String attributeName, String attributeId) throws WSClientTransportException;

    /**
     * Removes an attribute.
     * 
     * @param <Type>
     * @param userId
     * @param attribute
     * @throws WSClientTransportException
     */
    <Type> void removeAttribute(String userId, Attribute<Type> attribute) throws WSClientTransportException;
}

File: safe-online-sdk-dotnet/src/main/cs/ws/data/DataClient.cs
/*
 * SafeOnline project.
 * 
 * Copyright 2006-2008 	Lin.k N.V. All rights reserved.
 * Lin.k N.V. proprietary/confidential. Use is subject to license terms.
 */

using System;

namespace safe_online_sdk_dotnet
{
	public interface DataClient
	{
		void setAttributeValue(string userId, string attributeName, object attributeValue);
		
		/// <summary>
		/// Gives back the attribute value of an attribute.
		/// If attribute is missing a AttributeNotFoundException is thrown.
		/// null is returned if the attribute value is actually null.
		/// </summary>
		/// <param name="userId"></param>
		/// <param name="attributeName"></param>
		/// <returns>default(T), in case T is of type DateTime DateTime.MIN_VALUE is returned.</returns>
		/// <exception cref="RuntimeException">Invalid response received.</exception>
		/// <exception cref="RequestDeniedException">Data request denied.</exception>
		/// <exception cref="SubjectNotFoundException">Subject not found.</exception>
		T getAttributeValue<T>(string userId, string attributeName);

		/// <summary>
		/// Creates a new (empty) attribute for the given subject.
		/// </summary>
		/// <param name="userId"></param>
		/// <param name="attributeName"></param>
		/// <param name="objectValue">A string, DateTime, ... or an array of those types or 
		/// a Dictionary<string, object> in case of compound.</string,></param>
		/// <exception cref="RuntimeException">Creation failed.</exception>
		void createAttribute(string userId, string attributeName, object objectValue);
		
		/// <summary>
		/// Removes an attribute for the given subject.
		/// </summary>
		/// <param name="userId">the subject from which to remove the attribute</param>
		/// <param name="attributeName">the name of the attribute to be removed</param>
		/// <param name="attributeId">the optional attributeId in case of a compound attribute</param>
		/// <exception cref="RuntimeException">Delete failed.</exception>
		void removeAttribute(string userId, string attributeName, string attributeId);
		
	}
}
